service/javadsl/api/src/main/java/com/lightbend/lagom/javadsl/api/transport/TransportException.java

```diff
@@ -76,21 +76,27 @@ public ExceptionMessage exceptionMessage() {
     private static final Map<TransportErrorCode, BiFunction<TransportErrorCode, ExceptionMessage, TransportException>> BY_CODE_TRANSPORT_EXCEPTIONS;
 
     static {
+        // this map keeps a more strict relation between exception message and exception instances.
+        // Some exceptions reuse the same status code on certain transports so deserialization should
+        // try to reconstruct the exception by name first and fallback to reconstructing by code.
         Map<String, BiFunction<TransportErrorCode, ExceptionMessage, TransportException>> byName = new HashMap<>();
         byName.put(DeserializationException.class.getSimpleName(), DeserializationException::new);
-        byName.put(SerializationException.class.getSimpleName(), SerializationException::new);
-        byName.put(UnsupportedMediaType.class.getSimpleName(), UnsupportedMediaType::new);
-        byName.put(NotAcceptable.class.getSimpleName(), NotAcceptable::new);
+        byName.put(BadRequest.class.getSimpleName(), BadRequest::new);
+        byName.put(Forbidden.class.getSimpleName(), Forbidden::new);
         byName.put(PolicyViolation.class.getSimpleName(), PolicyViolation::new);
         byName.put(NotFound.class.getSimpleName(), NotFound::new);
-        byName.put(Forbidden.class.getSimpleName(), Forbidden::new);
+        byName.put(NotAcceptable.class.getSimpleName(), NotAcceptable::new);
+        byName.put(PayloadTooLarge.class.getSimpleName(), PayloadTooLarge::new);
+        byName.put(SerializationException.class.getSimpleName(), SerializationException::new);
+        byName.put(UnsupportedMediaType.class.getSimpleName(), UnsupportedMediaType::new);
```

Comment:
Here's where the actual error was: the default `ExceptionSerializer` uses `TransportException.fromCodeAndMessage` which tries to determine the exception by name and falls back to a lookup by status code. `BadRequest` and `PayloadTooLarge` were missing from the names list.

The changes is bigger than it should because also sorted the exception by status code value (400, 403, 404...). With that sorting convention it's easier to detect when something is missing.